### ISSUE TITLE
Performance Issues with RoBERTa Models  

### ISSUE THREAD

**Issue by Phil1108** [stanford-futuredata/ColBERT#12]
I am currently training a multilingual Model with your approach and with the bert-base-multilingual-uncased it works great.
Now I tried  switching to xlm-roberta-base (which in general is better pre-trained than the mBERT) but performance is far off. 
Both are trained on the same system with the same batch size. 

**Here a plot of the loss over training steps:**

![Colbert_losss](https://user-images.githubusercontent.com/39518904/100516308-3daa9b80-3183-11eb-8c66-d0c5f677ecf4.png)

Evaluation performance is very different as well: 
mBERT @ 32k Steps: **MRR@10 0.22**
XLM-RoBERTa @ 32k Steps: **MRR@10 0.07** 

As RoBERTa uses a BPE vocab i had to add unused tokens by hand and initalize the embedding for them randomly (transformers does that with mean=0 and std=0.02):
```       
        self.tokenizer = XLMRobertaTokenizer.from_pretrained('xlm-roberta-base')
        self.tokenizer.add_tokens(["[unused0]"])
        self.tokenizer.add_tokens(["[unused1]"])
        
        self.skiplist = {w: True for w in string.punctuation}

        self.bert = XLMRobertaModel(config)
        self.bert.resize_token_embeddings(len(self.tokenizer)) 
```
https://github.com/Phil1108/ColBERT/blob/b786e2e7ef1c13bac97a5f8a35aa9e5ff9f3425f/src/model.py#L24

Is the bad performance caused by the problem that RoBERTa doesn't include a next-sentence Prediction in its pre-training? Or is the ColBERT approach not transferable to BPE vocabs? 
Or is my way of adding unused tokens to the vocab causing undetected problems?

**Comment by okhat** [comment]
Hi Philipp,

I'm *not* surprised that RoBERTa is worse than BERT with ColBERT. As you say, I do suspect it may have to do with next-sentence prediction.

But I'm surprised at how large the difference is. Did you try tuning the learning rate? The default 3e-06 may be too small for RoBERTa.

Also, when you modify the tokenizer, do you do this in training only? Or is it applied for inference as well? You want to be sure the model is not using the wrong tokenizer later.

Lastly, when trying a new model you may want to avoid skipping punctuation just to be sure.

**Comment by okhat** [comment]
Seems resolved? Closing but feel free to re-open if you have further questions about this.

**Comment by Phil1108** [comment]
Hi, 
I have tried lots of things but the loss behaves weird. 

So tuning the learning rate seems to make it worse, even when introducing a learning rate with linear decay: 
![grafik](https://user-images.githubusercontent.com/39518904/100865328-7a360a00-3497-11eb-829b-86dbf9a8bb98.png)

This would support the proposal that RoBERTa doesn't work with the ColBERT approach due to the missing NSP training objective. 

However when trying to use ALBERT (which is trained with a sophisticated Sentence Order Objective) this leads to similar results as RoBERTa. 
![grafik](https://user-images.githubusercontent.com/39518904/100865660-e7499f80-3497-11eb-9e71-2ba2c1718f3c.png)

Also this paper which is i think very close to the ColBERT paper (https://www.aclweb.org/anthology/2020.coling-main.568.pdf) reports an increase in accuracy when using RoBERTa over BERT. (So the missing NSP doesn't seem to really influence these passage ranking tasks). 

One possible explenation would be that RoBERTa needs far more fine tuning steps (or larger batch sizes) to achieve the same result. So I recently switched to A100 GPUs, increased the BS from 32 to 150 and now I am waiting for the training and evaluation to finish.

**Comment by Phil1108** [comment]
> > Also, when you modify the tokenizer, do you do this in training only? Or is it applied for inference as well? You want to be sure the model is not using the wrong tokenizer later.
> 

No I explicitly checked that through printing out the input_ids. (This code is not yet uploaded in my fork because I do the inference and evaluation locally) 

Additionally I changed [unused0] & [unused1] tokens both to [UNK] and found out that it only had minor differences on the BERT performance. So it could hardly be a tokenizer/vocab problem, I guess.


> > Lastly, when trying a new model you may want to avoid skipping punctuation just to be sure.
> 
> I'm curious about your response to these.

I've done that here quick and dirty: https://github.com/Phil1108/ColBERT/blob/XLM-lr/src/model.py#L27
With skipping these, the performance became a little bit better, but not significantly. 

(Both answers here refer to the 32k Checkpoints not the 400k due to time limitations)

**Comment by okhat** [comment]
I see. I'd stick to the original ColBERT then, based on BERT and its multilingual variants.

I can't offer a lot of details about multilingual RoBERTa models without trying them myself.

**Comment by Phil1108** [comment]
Yeah I see. Multilingual BERTs work very well, even cross lingual. Loss of accuracy is not significantly in comparison with the english only. With the larger Batch Size on the A100s I think we can even beet the english-only model.

Sadly, there is still not a single open-sourced Large Multilingual BERT (It seems to me that Google wants to keep them private). That was the initial reason why I had to switch to XLM-RoBERTa (Large) Models because these are nearly the only available multilingual Large Models

**Comment by xuuuluuu** [comment]
@Phil1108 I am working on a similar idea on roBERTa. After adding the unused token to the vocab,

self.tokenizer.add_tokens(["[unused1]"])

the model's not working properly. Have you figured out where the problem is?

**Comment by Phil1108** [comment]
@xuuuluuu 
No I felt like I was stuck in a dead end there, so I stopped working on (XLM)-RoBERTa. 

My best guess is still that the missing next sentence prediction from RoBERTa is the main problem here (I think some people also discovered that as a reason why roberta is hard to use as a SBERT base model). 

I am not saying that it won't work at all, but it will probably need far more time to find a suitable configuration. 

You can see my code here: https://github.com/Phil1108/ColBERT/tree/XLM-lr

**Comment by xuuuluuu** [comment]
@Phil1108 
Thanks for the reply. My recent finding is that the roBERTa model performs worse when the added new token is critical to the task. It seems that the embedding of the newly added token cannot be properly learned. I am checking the reason...

**Comment by Phil1108** [comment]
@xuuuluuu 
Ah okay that could also be the reason, maybe it is the BPE that is not working in its intended way.

Even though it suprises me a little bit, because I have run experiments with BERT were I accidently assigned the two special tokens wrong so both became [unknown] tokens and it still produced better results than RoBERTa. 
So my overall conclusion was that the introduction of the tokens boost performance but BERT learns very well even without them. 

Have you tried to get completely rid of the special tokens?

**Comment by xuuuluuu** [comment]
@Phil1108 

Without adding the special tokens, the tokens become "[unknown]" to the RoBERTa model. As my model heavily relies on those tokens, it can be understood that the model performs badly. It is surprising that the BERT model performs better even with "[unknown]" than RoBERTa.

I found that the vocab of RoBERTa model has three unused tokens: "madeupword0000": 50261, "madeupword0001": 50262, "madeupword0002": 50263. Not sure if you can simply use these tokens for your case. For my case, I need much more special tokens.

**Comment by surajnair04** [comment]
@Phil1108 @xuuuluuu The issue arises due to the use of an incorrect model prefix in this line
https://github.com/Phil1108/ColBERT/blob/b786e2e7ef1c13bac97a5f8a35aa9e5ff9f3425f/src/model.py#L30 

For XLM-RoBERTa (or RoBERTa) models, the prefix should be roberta and not bert. Due to this mismatch, the model parameters are not loaded from the pretrained checkpoint and rather initialized from scratch. 
Here's a related issue: https://github.com/huggingface/transformers/issues/8407

**Comment by puppetm4st3r** [comment]
@Phil1108 do you resolve the issue? i'm looking for a multilanguaje model like colbert

**Comment by Phil1108** [comment]
@puppetm4st3r 
I have not trained an XLM RoBERTa with the solution provided above. 
But as you can see in my charts above I once trained a colbert based on multilingual BERT that worked quite fine. 
Have not uploaded it anywhere though, might do it when I have some time in the next couple of weeks. 

If its urgent just write me a dm.

**Comment by hiepxanh** [comment]
@Phil1108 my model is too bad, can you share your work? Maybe it can help me improve
